xrootd/streammanager/streammanager.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package streammanager // import "go-hep.org/x/hep/xrootd/streammanager"
+
+import (
+	"math"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+type ServerResponse struct {
+	Data  []byte
+	Error error
+}
+
+type dataSendChannel chan<- *ServerResponse
+type DataReceiveChannel <-chan *ServerResponse
+
+const streamIDPartSize = math.MaxUint8
+const streamIDPoolSize = streamIDPartSize * streamIDPartSize
+
+// StreamManager manages channels by their ids.
+// Basically, it's a map[StreamID] chan<-*ServerResponse
+// with methods to claim, free and pass data to specific channel by id.
+type StreamManager struct {
+	dataWaiters map[protocol.StreamID]dataSendChannel
+	freeIds     chan protocol.StreamID
+	mutex       sync.Mutex
+}
+
+// New creates new StreamManager
+func New() *StreamManager {
+	sm := StreamManager{
+		make(map[protocol.StreamID]dataSendChannel),
+		make(chan protocol.StreamID, streamIDPoolSize),
+		sync.Mutex{},
+	}
+
+	for firstByte := uint16(0); firstByte < streamIDPartSize; firstByte++ {
```

Comment:
wouldn't it be more efficient to allocate a global `[streamIDPoolSize]protocol.StreamID` array and launch a goroutine that would iterate over that.
the goroutine would send the current index to a non-buffered channel.
the stream-manager would draw from that non-buffered channel as needed.

this goroutine would need to be properly collected when we are done with the StreamManager.

Reply:
Can you please explain how can I collect goroutine?

Btw, when I make a goroutine like this:
```
go func() {
    i := 0
    for {
        sm.freeIDs <- i
        i = (i + 1) % streamIDPoolSize
    }
}()
```
where `freeIDs` is `make(chan int, streamIDPoolSize)`
it gives ~530 ns/op for Claim and Unclaim in benchmark comparing to  440 ns/op for current implementation.

Reply:
also, I am bit concerned with memory usage of this `make(chan protocol.StreamID, streamIDPoolSize)`
where: `streamIDPoolSize = 255*255 = 65025`
adding the size of each element of that buffered channel, we get: `65025 * 2 = 130050` bytes.

and that is per `StreamManager` value.

couldn't we find a way where we don't have to materialize these big structures?
(just working off simple integers `int` (or `int16`) and automatically create `protocol.StreamID` from/to these? a bit of `<<` and `>>` magic sounds like it would do.)

at the very least, we should mark all of these big structures with a clear `TODO` (+ an issue) so we don't forget about this.